src/node/txdownloadman_impl.cpp

```diff
@@ -0,0 +1,530 @@
+}
+CTransactionRef TxDownloadManager::GetTxToReconsider(NodeId nodeid)
+{
+    return m_impl->GetTxToReconsider(nodeid);
+}
+void TxDownloadManager::CheckIsEmpty() const
+{
+    return m_impl->CheckIsEmpty();
+}
+void TxDownloadManager::CheckIsEmpty(NodeId nodeid) const
+{
+    return m_impl->CheckIsEmpty(nodeid);
+}
+
+// TxDownloadManagerImpl
+void TxDownloadManagerImpl::ActiveTipChange()
+{
+    RecentRejectsFilter().reset();
+    RecentRejectsReconsiderableFilter().reset();
+}
+
+void TxDownloadManagerImpl::BlockConnected(const std::shared_ptr<const CBlock>& pblock)
+{
+    m_orphanage.EraseForBlock(*pblock);
+
+    for (const auto& ptx : pblock->vtx) {
+        RecentConfirmedTransactionsFilter().insert(ptx->GetHash().ToUint256());
+        if (ptx->HasWitness()) {
+            RecentConfirmedTransactionsFilter().insert(ptx->GetWitnessHash().ToUint256());
+        }
+        m_txrequest.ForgetTxHash(ptx->GetHash());
+        m_txrequest.ForgetTxHash(ptx->GetWitnessHash());
+    }
+}
+
+void TxDownloadManagerImpl::BlockDisconnected()
+{
+    // To avoid relay problems with transactions that were previously
+    // confirmed, clear our filter of recently confirmed transactions whenever
+    // there's a reorg.
+    // This means that in a 1-block reorg (where 1 block is disconnected and
+    // then another block reconnected), our filter will drop to having only one
+    // block's worth of transactions in it, but that should be fine, since
+    // presumably the most common case of relaying a confirmed transaction
+    // should be just after a new block containing it is found.
+    RecentConfirmedTransactionsFilter().reset();
+}
+
+bool TxDownloadManagerImpl::AlreadyHaveTx(const GenTxid& gtxid, bool include_reconsiderable)
+{
+    const uint256& hash = gtxid.GetHash();
+
+    if (gtxid.IsWtxid()) {
+        // Normal query by wtxid.
+        if (m_orphanage.HaveTx(Wtxid::FromUint256(hash))) return true;
+    } else {
+        // Never query by txid: it is possible that the transaction in the orphanage has the same
+        // txid but a different witness, which would give us a false positive result. If we decided
+        // not to request the transaction based on this result, an attacker could prevent us from
+        // downloading a transaction by intentionally creating a malleated version of it.  While
+        // only one (or none!) of these transactions can ultimately be confirmed, we have no way of
+        // discerning which one that is, so the orphanage can store multiple transactions with the
+        // same txid.
+        //
+        // While we won't query by txid, we can try to "guess" what the wtxid is based on the txid.
+        // A non-segwit transaction's txid == wtxid. Query this txid "casted" to a wtxid. This will
+        // help us find non-segwit transactions, saving bandwidth, and should have no false positives.
+        if (m_orphanage.HaveTx(Wtxid::FromUint256(hash))) return true;
+    }
```

Comment:
I don't understand the purpose of this, basically if orphanage has the hash transaction, we're returning true, regardless of whether IsWtxid or not.
```suggestion
    if (m_orphanage.HaveTx(Wtxid::FromUint256(hash))) return true;
```